packages/types/src/appearance.ts

```diff
@@ -0,0 +1,112 @@
+import * as CSS from 'csstype';
+
+import { OAuthProvider } from './oauth';
+
+type CSSProperties = CSS.PropertiesFallback<number | string>;
+type CSSPropertiesWithMultiValues = { [K in keyof CSSProperties]: CSSProperties[K] };
+type CSSPseudos = { [K in CSS.Pseudos as `&${K}`]?: CSSObject };
+
+interface CSSObject extends CSSPropertiesWithMultiValues, CSSPseudos {}
+type UserDefinedStyle = string | CSSObject;
+type ColorScale = Partial<Record<50 | 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900, string>>;
+type ColorOption = string | ColorScale;
+
+type LoadingState = 'loading';
+type DisabledState = 'disabled';
+type ErrorState = 'error';
+export type ElementState = LoadingState | DisabledState | ErrorState;
+
+/**
+ * A type that describes the states and the ids that we will combine
+ * in order to create all theming combinations
+ * If jsx exists, the element can also receive a typed function that returns a JSX.Element
+ */
+type ConfigOptions = { states: ElementState; ids: string; jsx: any };
+type WithOptions<Ids, States, Jsx> = { ids: Ids; states: States; jsx: Jsx };
+
+/**
+ * Create a type union of all state + id combinations
+ */
+export type StateSelectors<E extends string, S extends ElementState | undefined = never> = S extends never
+  ? never
+  : `${E}__${S}`;
+export type IdSelectors<E extends string, Id extends string | undefined = never> = Id extends never
+  ? never
+  : `${E}__${Id}`;
+type ElementPartsKeys<Name extends string, Opts extends ConfigOptions> =
+  | StateSelectors<Name, Opts['states']>
+  | IdSelectors<Name, Opts['ids']>
+  | StateSelectors<IdSelectors<Name, Opts['ids']>, Opts['states']>;
+type Selectors<RootElemName extends string, Opts extends ConfigOptions> =
+  | Partial<Record<RootElemName, UserDefinedStyle | Opts['jsx']>>
+  | Partial<Record<ElementPartsKeys<RootElemName, Opts>, UserDefinedStyle>>;
+
+export type ElementObjectKey<K extends string> = K extends `${infer Parent}-${infer Rest}`
+  ? `${Parent}${Capitalize<Rest>}`
+  : K;
+
+/**
+ * A map that describes the possible combinations we need to generate
+ * for each unique base element
+ */
+export type ElementsConfig = {
```

Comment:
Since you were interested, here's how the type generation happens cc: @colinclerk 
This file generates the public `Appearance` type using the `ElementsConfig` internal type as the single source of truth.

The generation of the actual runtime values our internal elements use happens here: packages/clerk-js/src/v4/customizables/elementDescriptors.ts 

When created, every `ElementDescriptor` object is bound to the `ElementsConfig` type. They are used to connect an element to a specific key of the appearance object. Since both the public `Appearance` type and the internal `ElementDescriptor` objects are using the `ElementsConfig` type, changing the one without updating the other would break - introducing accidental breaking changes to the customisation API should be pretty hard now

Reply:
Very nice!